Support Release Notes Generation

caithagoras opened the pull request:
The PR introduces a tool which formalize the release notes collection in the release process.
Address part of #13970.

# Release Notes Collection
To streamline the release process, we introduced a requirement (and PR template) to ask authors to write release notes in the PR description. During the release process, the person doing the release will be able to use this tool to collect and generate the release commit and PR, before polishing it as according to the 
[Release Notes Guideline](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines).

# What does the tool do
1. Switch to master branch. Fast-forward to upstream master.
2. Use `git log` to find the commits introduced in the new release.
3. Use [Github GraphQL API v4](https://developer.github.com/v4/) to fetch the PRs associated with the commits.
4. Parse the PR description, generate release notes, create release branch, modify local files, create release notes commit, and push to origin.
5. Create a release notes PR, with description populated. Release notes PR description contains 3 sections:
  a. A list of PRs with missing release notes, either the PR description do not contain release notes, or the release notes were malformed.
  b. A list of extracted release notes.
  c. A list of all commits introduced in the release.

# How to use the tool
To collect release notes, run the following command after cutting the release branch (bumping the snapshot version on the master branch):
```bash
curl -L -o /tmp/presto_release "https://oss.sonatype.org/service/local/artifact/maven/redirect?g=com.facebook.presto&a=presto-release&v=0.232-20200123.012813-4&r=snapshots&c=executable&e=jar"
chmod 755 /tmp/presto_release
/tmp/presto_release release-notes --github-user <GITHUB_USER> --github-access-token <GITHUB_ACCESS_TOKEN>
```

The changes have been tested end-to-end. Idea for unit testing are welcomed.

```
== RELEASE NOTES ==
Release Tool Changes
* Add a tool to collect and generate release notes (:pr:`13964`).
```

caithagoras commented:
@aweisberg Since @mayankgarg1990 is on PTO, could you help reviewing this PR? Thank you!

aweisberg commented on presto-release/src/main/java/com/facebook/presto/release/git/GitCommandsBinder.java:
```diff
@@ -0,0 +1,84 @@
+import static com.facebook.airlift.configuration.ConfigBinder.configBinder;
+import static java.util.Objects.requireNonNull;
+
+public class GitCommandsBinder
```
For people who don't know how dependency injection works this is hard to understand without looking at external resources.

A short explanation of what the binding step is would be helpful.

caithagoras replied:
Added the following comments to the class
```
   // To perform the binding: gitBinder(binder).bind(annotation, repositoryName)
```

aweisberg commented on presto-release/src/main/java/com/facebook/presto/release/git/PullRequest.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public int compareTo(PullRequest o)
+    {
+        return Integer.compare(id, o.id);
```
Strange that comapareTo only uses some of the fields?

caithagoras replied:
Removed the method `comparetTo`. I needed it to sort PR by PR# - changed to use `Collection.sort`

aweisberg commented on presto-release/src/main/java/com/facebook/presto/release/tasks/GenerateReleaseNotesCommand.java:
```diff
@@ -0,0 +1,106 @@
+{
+    private static final Logger log = Logger.get(GenerateReleaseNotesCommand.class);
+
+    @Option(name = "github-user", title = "Github username", required = true)
```
Is this the right convention for parameters names to not have - in front? I find it's useful for making it clear on the CLI the difference between the name of a parameter value (or sequence of values).

The convention I am used to is a single '-' in front of a single letter parameter name and '--' in front of the full word. And usually both would be available for each parameter.

caithagoras replied:
Fixed

aweisberg commented on presto-release/src/main/java/com/facebook/presto/release/tasks/GenerateReleaseNotesTask.java:
```diff
@@ -0,0 +1,443 @@
+        log.info("Release notes pull request created: %s", releaseNotesPullRequest.getUrl());
+    }
+
+    private Optional<List<ReleaseNoteItem>> extractReleaseNotes(PullRequest pullRequest)
```
It looks like some of this PR entry didn't get preserved https://github.com/prestodb/presto/pull/13604 All the lines starting with + got turned into a run-on sentence.

caithagoras replied:
Not sure what this comment means?

aweisberg replied:
If you run this tool it generates 
![image](https://user-images.githubusercontent.com/1047741/73377331-e13dcd80-428c-11ea-97be-ad22c457f0dc.png)
and
![image](https://user-images.githubusercontent.com/1047741/73377366-f1ee4380-428c-11ea-9990-782cf0672b88.png)
From that PR.

caithagoras replied:
This is the expected behavior since we never said we would support sub-bullet-point.

caithagoras commented on presto-release/src/main/java/com/facebook/presto/release/ForPresto.java:
```diff
@@ -0,0 +1,31 @@
+@Retention(RUNTIME)
+@Target({FIELD, PARAMETER, METHOD})
+@Qualifier
+public @interface ForPresto
```
It means "something related to Presto". Admitted, it is vague without any context. In this PR, there are 2 usages:
- `@ForPresto Git git`: a object to perform git operations on presto repo
- `@ForPresto LocalRepository repository`: represents the local git repo for Presto

Comparatively, we will also have `@ForPrestoFacebook` in a to-be-exist presto-facebook-release module.

We have 60+ of those annotations in presto, but none has comment - I think their exact meanings are better documented by usages in code rather than comment?

aweisberg replied:
I'll leave it up to you as to whether you want to add comment.

When doing code review I have an amount of effort I put into understanding what something does. Once I exceed that level of effort I start to think about whether a comment will save enough time for the people that come later. Code is write once read many so IMO the ROI is there.

There are some dependency injection related idioms here that still not very familiar with so it tripped me up enough to ask.

caithagoras commented on presto-release/src/main/java/com/facebook/presto/release/AbstractAnnotatedProvider.java:
```diff
@@ -0,0 +1,51 @@
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public abstract class AbstractAnnotatedProvider<T>
```
Comment added. This is similar to `com.facebook.presto.facebook.mysql.AbstractAnnotatedProvider`

caithagoras commented on presto-release/src/main/java/com/facebook/presto/release/git/GitActor.java:
```diff
@@ -0,0 +1,35 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class GitActor
```
This is how objects are defined in Github API:
Actor (Interface): https://developer.github.com/v4/interface/actor/
GitActor (Object): https://developer.github.com/v4/object/gitactor/

My understanding is that the "Git" in "GitActor" refers to git, the version control system, while `Actor` is more of a short form for `GithubActor`.

caithagoras commented on presto-release/src/main/java/com/facebook/presto/release/git/Actor.java:
```diff
@@ -0,0 +1,35 @@
+
+import static java.util.Objects.requireNonNull;
+
+public class Actor
```
This also explains the above comment.
# Differences
The author of a PR is an `Actor` - it has a login, but might not have a `name` unless it is a `User`.
The author of a commit is a `GitActor` - it has a name and an email, but may or may not has a Github account. (The commit author does not necessarily has to be the owner of the/a Github account)

caithagoras commented:
@aweisberg 
Added documentation to `presto-release/README.md`. Will add a website documentation in the future once the release tools are complete.

When there are missing release notes, authors/committers are encouraged to comment on the release notes PR and checkbox the item, so that the release oncall can incorporate them into the PR. Usually, there are only a few missing items, and the release oncall need to polish the notes, so it might just be easier to do it manually. There are potentials to strength the tool to make it smarter, but I think it is good enough for now.

caithagoras commented:
@aweisberg Added `TestGenerateReleaseNotesTask`.

caithagoras commented:
- Updated `extractReleaseNote` logic to fix some bugs.
- Fixed an issue in `GithubAction.listCommits` where the latest commit is not included in the release notes.

caithagoras commented:
@aweisberg Ready for review again. Thanks!

mayankgarg1990 commented on presto-release/src/main/java/com/facebook/presto/release/PrestoReleaseService.java:
```diff
@@ -0,0 +1,36 @@
+
+    public static void main(String[] args)
+    {
+        Cli<Runnable> parser = Cli.<Runnable>builder("verifier")
```
`release`

mayankgarg1990 commented on presto-release/src/main/java/com/facebook/presto/release/PrestoReleaseService.java:
```diff
@@ -0,0 +1,36 @@
+    public static void main(String[] args)
+    {
+        Cli<Runnable> parser = Cli.<Runnable>builder("verifier")
+                .withDescription("Presto Verifier")
```
`Presto Release`

aweisberg commented:
When we do reviews can you add your changes as commits and then rebase to squash and merge them after review is complete? It's makes it easier to see how you respond to comments and how things are evolving since I last reviewed compared to seeing the whole thing force pushed.

caithagoras commented:
@aweisberg Sure, will do. Thanks!

caithagoras commented:
@mayankgarg1990 Good idea. I think we should, and the library seems to have the things we need as I scanned through it. But since there are no functional changes, maybe we can iterate on it later.

caithagoras commented:
@mayankgarg1990 I looked into jgit and found some issue with using it. When working with ssh, jgit needs to parse the private key file and it does not support the RFC4716 format.

> -m key_format
>             Specify a key format for the -i (import) or -e (export) conversion options.  The supported key formats are: ``RFC4716'' (RFC 4716/SSH2 public or private key), ``PKCS8'' (PEM PKCS8 public key) or ``PEM''
>             (PEM public key).  The default conversion format is ``RFC4716''.  Setting a format of ``PEM'' when generating or updating a supported private key type will cause the key to be stored in the legacy PEM
>             private key format.

tdcmeehan commented:
I came across this PR.  I just skimmed and have a couple of high-level questions.

1) Does this need to exist inside the main PrestoDB repo?  Can it exist in a stand-alone repository?  The Presto build is already long and the codebase already vast, it would be better to put this somewhere independent if there's no code sharing.
2) The lines of code are quite hefty and we'll need to maintain this is as part of the mainline Presto build.  Could it have been written more concisely with some 3rd party libraries, or even with a different language like Python?

caithagoras commented on presto-release/README.md:
````diff
@@ -0,0 +1,13 @@
+## Generate Release Notes
+To collect and generate release notes, run the following commands in the presto repo.
+```
+curl -L -o /tmp/presto_release "https://oss.sonatype.org/service/local/artifact/maven/redirect?g=com.facebook.presto&a=presto-release&v=0.232-20200123.012813-4&r=snapshots&c=executable&e=jar"
````
Updated. We can do `v=LATEST` to fetch the latest snapshot version. When there is a release of this project, we can change it to point to the latest release version.

caithagoras commented on presto-release/NOTICE:
```diff
@@ -0,0 +1,2836 @@
+THE FOLLOWING IS SOFTWARE LICENSED BY THIRD PARTIES UNDER OPEN SOURCE LICENSES THAT MAY BE USED BY THIS PRODUCT.
```
I'm moving `NOTICE` and `README.md` to the top level. They're specified `src/main/assembly/presto-release-tools.xml`

caithagoras commented:
Superseded by https://github.com/prestodb/presto-release-tools/pull/2
